Add support for ppc64le architecture

hkshaw1990 opened the pull request:
Sets the GOARCH according to the current architecture. Modifies the
yaml.tmpl files to add images and tags for zipkin, prometheus,
statsd-exporter, stackdriver/zipkin-collector, etc. Adds 3 dockerfiles
for mixer_debug, servicegraph_debug and graphana, since, no debug
images for ppc64le in istio-testing/ubuntu_xenial_debug. Modifies the
tools/istio-docker.mk accordingly. The tests with minikube need to
be disabled for ppc64le, since minikube binary is not publically
available on github for ppc64le.

istio-testing commented:
Hi @hkshaw1990. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should join the org to skip this step.

I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).

<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>

hkshaw1990 commented:
I signed it!

mattdelco commented on Makefile:
```diff
@@ -54,7 +54,13 @@ ifeq ($(GO),)
+ifeq ($(MACH), ppc64le)
+   export GOARCH ?= ppc64le
+else
+   export GOARCH ?= amd64
```
For Linux a MACH of "x86_64" would need to be set to "amd64".  I didn't check what a Mac reports.

ldemailly replied:
my mac says:
```
$ uname -m
x86_64
```

hkshaw1990 replied:
@mattdelco so, this change should be fine, right? As @ldemailly reported, that a mac shows `x86_64` and I think is consistent with the x86 servers too. The `GOARCH` will only be set to `ppc64le` only if `MACH==ppc64le` and if `GOARCH` wasn't set to anything else before. Or, do we need to set that specifically for `MACH==x86_64`?

mattdelco replied:
I suppose it's fine.  I think my original mindset was that the 'else' case was "export GOARCH ?= $(MACH)", which would be a problem, but as you point out that's not how it was implemented.  I suppose it could be:

<pre>ifeq ($(MACH), ppc64le)
   export GOARCH ?= ppc64le
else ifeq ($(MACH), x86_64)
   export GOARCH ?= amd64
else
   export GOARCH ?= $(MACH)
endif</pre>

but I don't have a preference--it might depend on what arm reports (it's the only other platform that comes to mind offhand that's still in production and has decent popularity).

hkshaw1990 replied:
Yeah. I guess, I can add the `amd64` bits specifically.

mattdelco commented on addons/servicegraph/docker/Dockerfile.servicegraph_debug_ppc64le:
```diff
@@ -0,0 +1,9 @@
+FROM ubuntu:xenial
```
Rather than adding a new docker file an alternative is to make the existing file more customizable.  E.g. the original file would have something like:
<pre>
- FROM gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
+ UBUNTU_DEBUG_IMAGE=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
+ FROM ${UBUNTU_DEBUG_IMAGE}
</pre>
and then the Makefile would have something like:
<pre>
ifeq ($(MACH), ppc64le)
DOCKER_OPT+=--build-arg UBUNTU_DEBUG_IMAGE=ubuntu:xenial
endif
</pre>
and then add a $DOCKER_OPT after "docker build" in the DOCKER_RULE.

This proposal isn't perfect since the "FROM" for Linux vs PPC would live in two different places but it would help to avoid the duplication of files and reduce the severity of the changes in the makefile.  Since an actual debug image of xenial isn't being used I'm not sure if some one argue that it's be better to just not make the debug images part of a build on ppc.

I'm not sure what to do about grafana since it seems to differ more than just the from line.

ldemailly replied:
I'll show my lack of docker knowledge but isn't ubuntu:xenial a linux x86 image ?
(oh do you make non x86 images in docker?)

hkshaw1990 replied:
> Rather than adding a new docker file an alternative is to make the existing file more customizable.

@mattdelco yeah, this makes more sense. Will change this to a customizable base image. grafana has the same case too, i.e., it just differs w.r.t the base image.

>  isn't ubuntu:xenial a linux x86 image ? oh do you make non x86 images in docker?

@ldemailly No and yes. docker.io registry supports multi-arch manifest for some of the images. 
`ubuntu:xenial` is one of them. So, it basically depends from which architecture we are pulling the image from and will pull the appropriate image.

mattdelco commented on install/updateVersion.sh:
```diff
@@ -64,6 +69,10 @@ while getopts :i:n:p:x:c:a:h:r:P:d:D:m: arg; do
     P) PILOT_DEBIAN_URL="${OPTARG}";;
     d) DEST_DIR="${OPTARG}";;
     D) PROXY_DEBUG="${OPTARG}";;
+    s) STATSD_HUB_TAG="${OPTARG}";;
```
As I read things it seems like the original/default hub/tag for these 4 needs to be added to istio.VERSION as part of this change.

hkshaw1990 replied:
Right. Will add them here.

costinm commented on addons/grafana/Dockerfile.grafana_ppc64le:
```diff
@@ -0,0 +1,19 @@
+# starts graphana with Prometheus Datasource pre configured
```
I am not sure this will scale. 

Would it be possible to change the makefile to do a small sed on the Dockerfile, and replace
'from' with a custom version ?

This will allow future arm, mips, etc - and may also allow using a different, custom base

hkshaw1990 replied:
Agreed. It should be possible to do a sed on the Dockerfile. Will make that change.

mattdelco replied:
I think the sed approach adds a lot more complexity than the approach I suggested (over in Dockerfile.servicegraph_debug_ppc64le) of using a ARG variable in the from statement for the Dockerfile.  I've got a more concrete example here:

https://github.com/istio/istio/compare/master...mattdelco:master16

In that change only the changes to the Dockerfiles and istio-docker.mk are relevant.  That change is for a different use case of trying to consume a local proxy on an opt-in basis.  This change seems to be headed in the direction of always having the Makefile specify the value for the ARG, and that case should work fine too (the default value for the ARG in the Dockerfile can be omitted).  The main downside I noticed in my approach is that the same DOCKER_RULE is used for all docker images, and "docker build" logs a messages if the value of a --build-arg wasn't used in a Dockerfile (there's probably ways to address this if it annoys people).

hkshaw1990 replied:
> I think the sed approach adds a lot more complexity than the approach I suggested (over in Dockerfile.servicegraph_debug_ppc64le) of using a ARG variable in the from statement for the Dockerfile.

@mattdelco Well, the sed approach certainly adds more lines. I have added a commit which uses the sed approach. I am fine with taking the other approach (i.e., the ARG approach).

> The main downside I noticed in my approach is that the same DOCKER_RULE is used for all docker images, and "docker build" logs a messages if the value of a --build-arg wasn't used in a Dockerfile (there's probably ways to address this if it annoys people).

Well, if it annoys people, can't we make `DOCKER_OPT` "" (empty) [here](https://github.com/istio/istio/compare/master...mattdelco:master16) based on whether it should be used? Or else, we can have two separate `DOCKER_RULE`s, one with the arg and other without?

mattdelco replied:
I'm thinking that the customizing of FROM might increase over time for various reasons, partly because there's 3 you want to customize, and I had a different reason for another (though a recent change made mine use case morph into a different kind of problem), so more are conceivable.

Regardless, I've been thinking about ways to appropriately filter/prune the DOCKER_OPT for each docker image.  My first line of thinking involved using "$(filter" to pick out options that contained the target name (and then possibly use "$(subst" to prune off the part that contained the target name)), but now I'm thinking that a more straightforward option is to just use $(DOCKER_OPT$(suffix $@)) rather than a plain $(DOCKER_OPT) in DOCKER_RULE.  Then opts can be defined on a per-image basis, e.g.:

<pre>ifeq ($(MACH), ppc64le)
DOCKER_OPT.mixer_debug:=--build-arg MIXER_DEBUG_BASE=ubuntu:xenial
DOCKER_OPT.servicegraph_debug:=--build-arg SERVICEGRAPH_DEBUG_BASE=ubuntu:xenial
DOCKER_OPT.grafana:=--build-arg GRAFANA_BASE=ibmcom/grafana-ppc64le:4.4.3
else
DOCKER_OPT.mixer_debug:=--build-arg MIXER_DEBUG_BASE=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
DOCKER_OPT.servicegraph_debug:=--build-arg SERVICEGRAPH_DEBUG_BASE=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
DOCKER_OPT.grafana:=--build-arg GRAFANA_BASE=grafana/grafana:4.1.2
endif
</pre>
and then using mixer debug as an example its Dockerfile would have:
<pre>
ARG MIXER_DEBUG_BASE
FROM ${MIXER_DEBUG_BASE}
</pre>
As a test of the approach I made a standalone Makefile like this:
<pre>
HUB:=hub
TAG:=tag
SECURITY_DOCKER:=docker.istio-ca docker.istio-ca-test docker.node-agent docker.node-agent-test docker.multicluster-ca

default: $(SECURITY_DOCKER)

DOCKER_OPT.istio-ca:=--build-arg OPT_CA=ca:latest
DOCKER_OPT.istio-ca-test:=--build-arg OPT_CA_TEST=catest:latest
DOCKER_OPT.node-agent:=--build-arg OPT_NODE_AGENT=nodeagent:latest
DOCKER_OPT.node-agent-test:=--build-arg OPT_NODE_AGENT_TEST=nodeagenttest:latest

$(SECURITY_DOCKER):
	$(DOCKER_RULE)

DOCKER_RULE=@echo "docker build $(DOCKER_OPT$(suffix $@)) -t $(HUB)/$(subst docker.,,$@):$(TAG) -f Dockerfile$(suffix $@) ."
</pre>
and if I ran "make default" it output the commands with the desired customization:

<pre>
docker build --build-arg OPT_CA=ca:latest -t hub/istio-ca:tag -f Dockerfile.istio-ca .
docker build --build-arg OPT_CA_TEST=catest:latest -t hub/istio-ca-test:tag -f Dockerfile.istio-ca-test .
docker build --build-arg OPT_NODE_AGENT=nodeagent:latest -t hub/node-agent:tag -f Dockerfile.node-agent .
docker build --build-arg OPT_NODE_AGENT_TEST=nodeagenttest:latest -t hub/node-agent-test:tag -f Dockerfile.node-agent-test .
docker build -t hub/multicluster-ca:tag -f Dockerfile.multicluster-ca .
</pre>

I did simplify my initial example above, and I wouldn't be surprised if people desired that the image names be put on simpler standalone lines so that the update bots have less complex lines to parse, e.g.:

<pre>ifeq ($(MACH), ppc64le)
MIXER_DEBUG_BASE:=ubuntu:xenial
SERVICEGRAPH_DEBUG_BASE:=ubuntu:xenial
GRAFANA_BASE:=ibmcom/grafana-ppc64le:4.4.3
else
MIXER_DEBUG_BASE:=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
SERVICEGRAPH_DEBUG_BASE:=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
GRAFANA_BASE:=grafana/grafana:4.1.2
endif

DOCKER_OPT.mixer_debug:=--build-arg MIXER_DEBUG_BASE=$(MIXER_DEBUG_BASE)
DOCKER_OPT.servicegraph_debug:=--build-arg SERVICEGRAPH_DEBUG_BASE=$(SERVICEGRAPH_DEBUG_BASE)
DOCKER_OPT.grafana:=--build-arg GRAFANA_BASE=$(GRAFANA_BASE)
</pre>

hkshaw1990 replied:
@mattdelco thanks for the explanation. It does seem a more simple approach comparatively. I have added a [commit](https://github.com/istio/istio/pull/3684/commits/b597827b6600a062b474788850f0817d6aa191fe) here which implements the same.

costinm commented on bin/testEnvRootMinikube.sh:
```diff
@@ -2,6 +2,11 @@
+    exit 1
+fi
+
 if [ ! -f /usr/local/bin/minikube ]; then
```
Don't you want to also cross-compile (as we do for mac and arm ? )

hkshaw1990 replied:
I am assuming you are referring to this (`.circrleci/config.yml`):
```            make build GOOS=darwin KUBECONFIG=/go/src/istio.io/istio/.circleci/config
            make build GOOS=linux GOARCH=arm KUBECONFIG=/go/src/istio.io/istio/.circleci/config
```
The problem with this is, circleci doesn't support ppc64le hardware yet. We would definitely want to test istio with kubernetes on ppc64le. Is there any alternative approach to do that?

hkshaw1990 replied:
@costinm never mind. I just realised its just cross-compilation anyway :) , yes, we can add that for ppc64le as well. I will remove this change and try out cross-compiling for ppc64le as well.

costinm commented:
Regarding testing and possibly docker image building on circle: if you can
'donate' a small k8s cluster on ppc, an admin  can add the keys and circle
could run some of tests. Same for the docker image building - will likely
need an explicit config for remote docker.

I would like to eventually do the same for arm.

On Feb 23, 2018 05:52, "Hemant Kumar" <notifications@github.com> wrote:

*@hkshaw1990* commented on this pull request.
------------------------------

In bin/testEnvRootMinikube.sh
<https://github.com/istio/istio/pull/3684#discussion_r170255606>:

> @@ -2,6 +2,11 @@

 export K8S_VER=${K8S_VER:-v1.9.2}

+if [ "$ARCH" == "ppc64le" ]; then
+    echo "Minikube tests not supported on ppc64le, exiting..."
+    exit 1
+fi
+
 if [ ! -f /usr/local/bin/minikube ]; then

@costinm <https://github.com/costinm> never mind. I just realised its just
cross-compilation anyway :) , yes, we can add that for ppc64le as well. I
will remove this change and try out cross-compiling for ppc64le as well.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/istio/istio/pull/3684#discussion_r170255606>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAFI6q8JRY5FvgojXXqDks9ync-klfklks5tXsKNgaJpZM4SPa68>
.

hkshaw1990 commented:
>Regarding testing and possibly docker image building on circle: if you can
'donate' a small k8s cluster on ppc, an admin  can add the keys and circle
could run some of tests. Same for the docker image building - will likely
need an explicit config for remote docker.

Sure, makes sense. Once I have the setup, I will get back to you.

mattdelco commented on install/updateVersion.sh:
```diff
@@ -218,6 +247,14 @@ export FORTIO_HUB="${FORTIO_HUB}"
+export PROM_TAG="${PROM_TAG}"
+export ZIPKIN_HUB="${ZIPKIN_HUB}"
+export ZIPKIN_TAG="${ZIPKIN_TAG}"
+export ZIPKIN_COLLECTOR_HUB="${ZIPKIN_HUB}"
```
<pre>
-export ZIPKIN_COLLECTOR_HUB="${ZIPKIN_HUB}"
-export ZIPKIN_COLLECTOR_TAG="${ZIPKIN_TAG}"
+export ZIPKIN_COLLECTOR_HUB="${ZIPKIN_COLLECTOR_HUB}"
+export ZIPKIN_COLLECTOR_TAG="${ZIPKIN_COLLECTOR_TAG}"
</pre>

hkshaw1990 replied:
Sorry. Yeah, thats sounds right.

mattdelco commented on bin/gobuild.sh:
```diff
@@ -31,8 +31,14 @@ BUILDPATH=${3:?"path to build"}
 GOOS=${GOOS:-linux}
 GOARCH=${GOARCH:-amd64}
+if [ "$ARCH" == "ppc64le" ]; then
+    GOARCH=$ARCH
```
If I read this correctly, then someone running on a PPC system is forced to compile for PPC (i.e., cross-compiling isn't possible).  I'm thinking it'd be more appropriate to have something like the following (though there's probably a more concise way to get the same result):

<pre>
if [ "$ARCH" == "ppc64le" ]; then
    DEFAULT_ARCH=$ARCH
else
    DEFAULT_ARCH=amd64
endif

GOARCH=${GOARCH:-${DEFAULT_ARCH}}
</pre>

hkshaw1990 replied:
This suggestion makes sense. Have added the required changes.

mattdelco commented on install/updateVersion.sh:
```diff
@@ -291,6 +335,9 @@ function update_istio_addons() {
   execute_sed "s|image: {MIXER_HUB}/\(.*\):{MIXER_TAG}|image: ${MIXER_HUB}/\1:${MIXER_TAG}|" grafana.yaml.tmpl
   execute_sed "s|image: {MIXER_HUB}/\(.*\):{MIXER_TAG}|image: ${MIXER_HUB}/\1:${MIXER_TAG}|" servicegraph.yaml.tmpl
+  execute_sed "s|image: {PROM_HUB}/\(.*\):{PROM_TAG}|image: ${PROM_HUB}/\1:${PROM_TAG}|" prometheus.yaml.tmpl
+ execute_sed "s|image: {ZIPKIN_HUB}/\(.*\):{ZIPKIN_TAG}|image: ${ZIPKIN_HUB}/\1:${ZIPKIN_TAG}|" zipkin.yaml.tmpl
```
The indentation doesn't look consistent here, as if the two lines for zipkin are missing a space.

hkshaw1990 replied:
Fixed it.

mattdelco commented on istio.VERSION:
```diff
@@ -15,3 +15,11 @@ export FORTIO_HUB="docker.io/istio"
 export HYPERKUBE_HUB="quay.io/coreos/hyperkube"
 export HYPERKUBE_TAG="v1.7.6_coreos.0"
+export STATSD_HUB="prom/statsd-exporter"
+export STATSD_TAG="v0.5.0"
```
It looks like some of the original cases were using "latest" while others were using "v0.5.0".  I have no idea on which is more appropriate but I'll just point it out in case another reviewer has more knowledgeable thoughts on this.

hkshaw1990 replied:
Yeah, this looked confusing to me. If someone can clarify, I will add the right tag.

mattdelco commented on tools/istio-docker.mk:
```diff
@@ -37,6 +37,20 @@ GRAFANA_FILES:=addons/grafana/start.sh \
                addons/grafana/pilot-dashboard.json \
                addons/grafana/import_dashboard.sh
 
+ifeq ($(MACH), ppc64le)
```
Offhand I'm thinking that GOARCH is the more appropriate variable to test since we'd care about what platform we're compiling for rather than the platform we're compiling on.

hkshaw1990 replied:
@mattdelco from what I understand, this file is not meant for building go binaries. Its for building the docker images. And I guess, you would need a ppc64le platform for building the docker images. So, if we use GOARCH here, it won't make much sense, since, docker build won't proceed if its not able to pull the required images on a different platform. Please correct me if I am wrong.

hkshaw1990 replied:
> If docker can't handle cross compiles then I suppose all that really could be done is check that the GOARCH is unset or consistent with MACH (i.e., verify that docker is being asked to bundle go binaries that were compiled for the current/native platform), or just punt on this and leave open the possibility of the uncommon case of cross-compile users shooting themselves in the foot.

Well, it doesn't seem like a good idea to leave it just like that. I'll put a check as per your suggestion, which can be something like:
```
if $GOARCH != $MACH
    echo "Some error msg explaining docker images can't be built for cross platform building..."
endif
```
How does that sound?

mattdelco replied:
A check sounds fine, just don't forget to account for the case where GOARCH=amd64 and MACH=x86_64.  I suppose one way to keep the comparison in istio-docker.mk straightforward and simplify the diff in the Makefile is to try something like:

<pre>
MACH ?= $(shell uname -m)
ifeq ($(MACH),x86_64)
MACH := amd64
endif

export GOARCH ?= $(MACH)
</pre>

hkshaw1990 replied:
Cool, thanks! Added this.

hkshaw1990 commented on istio.VERSION:
```diff
@@ -15,3 +15,11 @@ export FORTIO_HUB="docker.io/istio"
+export PROM_TAG="v2.0.0"
+export ZIPKIN_HUB="docker.io/openzipkin/zipkin"
+export ZIPKIN_TAG="latest"
+export ZIPKIN_COLLECTOR_HUB="docker.io/openzipkin/zipkin"
```
Yeah, my mistake. Fixed it to `gcr.io/stackdriver-trace-docker/zipkin-collector`. Thanks!

mattdelco commented:
Offhand I don't have anything more to add, but we'll see how things look after you have to rebase to pick up other changes (e.g., istio.VERSION went away and the defaults for the hub/tag are now set in install/updateVersion.sh).

hkshaw1990 commented:
> Offhand I don't have anything more to add, but we'll see how things look after you have to rebase to pick up other changes (e.g., istio.VERSION went away and the defaults for the hub/tag are now set in install/updateVersion.sh).

I have pushed the changes after rebasing. Few points to note:
- As mentioned already, `istio.VERSION` went away, the default versions are now in `install/updateVersion.sh`.
- The `values.yaml.tmpl` went away too. This poses some issues for other architectures. The `values.yaml` contains the repositories for `prometheus`, `statsd` and `zipkin`. All of those are x86 images. Is there some way we can customise the `repository` and `tag` values for other platforms as well? Or else, can we move back to the template way?

mattdelco commented:
I'm not familiar with helm (nor the yalm files for that matter).  I'm not sure if the expectation for values.yaml is based around --set, or using another values-ppc.yaml (akin to values-istiotest.yaml) to override settings in values.yaml, or something else.  @costinm probably has thoughts on this.

sdake commented:
If overrides are desired, there are two techniques.  A -`-set` will override individual values on the command line, and a `-f <file>` will override only those values in the override file.  For more details review the Helm documentation here: https://github.com/kubernetes/helm/blob/master/docs/chart_template_guide/values_files.md

repository and tag can be overridden by `helm template istio --set globals.hub=${HUB} --set globals.tag=${TAG} > istio-ppc.yaml`.  Perhaps this can be automated in the build system.

hkshaw1990 commented:
@sdake @mattdelco So, when I tried to generate  `istio.yaml` based on `--set` with custom global.hub and global.tag, it sets the right hub and tag (whatever is provided on the command line). But, for other containers like statsd-exporter :
```
helm template istio --set prometheusStatsdExporter.repository=ibmcom/statsd-exporter > istio-ppc.yaml
```
its not setting the right `image` from `install/kubernetes/helm/istio/charts/mixer/templates/deployment.yaml`:
```
      containers:
        - name: statsd-to-prometheus
          image: "{{ .Values.prometheusStatsdExporter.repository }}:{{ .Values.prometheusStatsdExporter.tag}}"
          imagePullPolicy: {{ .Values.prometheusStatsdExporter.imagePullPolicy }}
```
The second case is where we would like to have a different image than the default one. It could be mostly because, I haven't yet understood the mechanism of values.yaml with `--set`.

> Perhaps this can be automated in the build system.

Assuming that the first case is resolved, how do we integrate this into the build system? Shall we just add a `helm template ...` command to `updateVersion.sh` depending on the ARCH?
Thanks for the help!

sdake commented:
Related PR: https://github.com/istio/istio/pull/3856

The above PR provides automatic node affinity by CPU type for multiarch images.

@hkshaw1990 note shortly @mandarjog will be removing the .tmpl files and replacing with helm template.

sdake commented:
@costinm note @hrw did a full port of OpenStack containers to arm and is an ARM arch expert.  He may be able to put you in touch with folks that can deliver on your desire here: https://github.com/istio/istio/pull/3684#issuecomment-368052986

sdake commented:
@hkshaw1990 try `helm template --set mixer.prometheusStatsdExporter.repository=ibmcom/statsd-exporter > istio-ppc.yaml`

We should probably move all image hub/tags to global (as a separate PR) as the current organization made sense when we had separate hub/tag overrides available.  @costinm unified nearly all of the HUB/TAG overrides into one variable, however, there were some that were left untouched (like statsd as you mention) as they come from other upstreams.

hkshaw1990 commented:
> helm template --set mixer.prometheusStatsdExporter.repository=ibmcom/statsd-exporter > istio-ppc.yaml

Thanks, override works. But again, until a we have a best possible solution to the build process for the helm charts for ppc64le, we can park this for now.

Please let me know if you have any other comments on this PR.

sdake commented:
@hkshaw1990 related to parking the helm aspect of things, the `*.tmpl` files have been removed in master, and all istio.yaml templates are produced via helm --template.  That work has been taking place for several weeks, which is why I suggested implementing the Helm work.

As is, this patch will need a rebase on master and need a Helm based implementation.

Reach out if you need some guidance.

Cheers
-steve

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: 
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **geeknoid**

Assign the PR to them by writing `/assign @geeknoid` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["geeknoid"]} -->

hkshaw1990 commented:
Hey @sdake 
> related to parking the helm aspect of things, the *.tmpl files have been removed in master, and all istio.yaml templates are produced via helm --template. That work has been taking place for several weeks, which is why I suggested implementing the Helm work.

Got it. One of the reasons why I suggested to park the helm based implementation for `ppc64le` was because of the right way to approach the problem. We would need a solution where the same helm charts can be shipped for `amd64` and `ppc64le`. And, the values (repository) for some of the docker images (statsd, prometheus and zipkin) should be determined in runtime without user intervention. But upon going through the different methods, I wasn't able to find a way to do that. I have added a [commit](https://github.com/istio/istio/pull/3684/commits/8c8e0276f0953820143d70e9e9cc31b304e2438d) which, based on the user provided value of `platform`, will set the right image repositories for `statsd`, `prometheus` and `zipkin`. Please let me know if it looks fine.
I guess, for now, we can use this (user provided) approach to figure out the appropriate images, until an arch based solution crops up in `helm`.

> As is, this patch will need a rebase on master and need a Helm based implementation.

Yeah, rebased it and added a helm based implementation.

Thanks!

sdake commented:
@hkshaw1990 after seeing the PR and your comment, I understand why you wanted to park the helm method.   I think this solution is workable but needs a better resolution prior to 1.0.  As is, LGTM.

The gate tests are failing WRT circleci.  Unfortunately I can't re-run the failing gates in circleci (I can only do this on my own PRs, not others).  The only way I've found to rerun the circleci tests in this case is to rebase.  @costinm may have permissions to get the gates in a passing state so this will merge.

Cheers
-steve

sdake commented:
/ok-to-test

gyliu513 commented on install/kubernetes/helm/istio/charts/mixer/templates/statsdtoprom.yaml:
```diff
@@ -50,7 +50,11 @@ spec:
       - name: {{ $statsdname }}
-        image: "{{ $.Values.prometheusStatsdExporter.repository }}:{{ $.Values.prometheusStatsdExporter.tag}}"
+    {{- if eq $.Values.global.platform "amd64" }}
+        image: "{{ $.Values.prometheusStatsdExporter.repository_ppc64le }}:{{ $.Values.prometheusStatsdExporter.tag}}"
```
How about use multi-arch docker image? Thus we can use one image name for different platforms.

hkshaw1990 replied:
We want to use a multi-arch docker image but prom/statsd-exporter doesn't have multi-arch manifest support yet. That's why two different repositories for `amd64` and `ppc64le`.

gyliu513 replied:
Can we write the manifest to do this?

hkshaw1990 replied:
@gyliu513 the problem is we don't have any control over those registries. 2-3 months back we did suggest to add dockerfiles for `ppc64le` to prometheus, but its been pending since. That's the reason we didn't pursue the manifest route for these images.

sdake replied:
@hkshaw1990 I'm not sure what the process is for adding Dockerfiles for different arches.  Can you provide a link to the Dockerfile PR that has been pending?

@ldemailly @costinm would you mind weighing in on this specific point (do we want to add these Dockerfiles to the istio registries) so we can move this PR along?

gyliu513 replied:
@hkshaw1990 can you please show me the issue that we opened in prometheus community?

hkshaw1990 replied:
@gyliu513 yeah, this is a link to one of our PRs :
https://github.com/prometheus/prometheus/pull/2693

clnperez replied:
If the istio project wanted to host manifest list that just points to those two, that would be doable. You'd just have a little extra overhead of keeping that list current, and I donno if it's worth it to save two lines of yaml.

clnperez replied:
also -- is this if condition flipped? i think the action has the wrong arch on this one.

hkshaw1990 replied:
@clnperez true, missed that. Will add a fix with the next rebase. Thanks!

hkshaw1990 commented:
@sdake 
> Also would you mind running through the helm linter?

Sure, I will do that.

Also, we thought a PR for istio/proxy and envoyproxy/envoy would be ready by now but we are consistently hitting new issues on ppc64le which we are trying to fix (see: [discussion](https://groups.google.com/forum/#!searchin/envoy-dev/christy%7Csort:date/envoy-dev/NkXq5u2COCE/M4FxONJeAgAJ)). So, the envoy PRs are not ready yet. We have been using an older version of envoy to test on our system with istio. Shall we still move ahead with this PR or wait till we get to the envoy PRs?

hkshaw1990 commented:
Hey @gyliu513,

> Can you please split this PR to two PRs?
1. Update Makefile and Dockerfile to support ppc64le.
2. Update helm chart.

Yeah, sure. But, is there any specific reason to do that since, the 2nd PR would be dependent on the 1st anyway?

gyliu513 commented:
> Yeah, sure. But, is there any specific reason to do that, since, the 2nd PR would be dependent on the 1st anyway?

@hkshaw1990 This can make the PR clear. Also, I think that the two PRs are not tightly coupled, as for now, I can build my own ppc64le image, and there are also some checking logic for ppc64le and x86.

sdake commented:
@hkshaw1990 I wouldn't block on proxy for ppc64.  Prior to 1.0, we do need to sanitize the helm `values.yaml` as it is an API and should be stable.  I'd ask that once you do get the images built for proxy sorted out, you refactor some of this work in the PR to remove the particular problem.

sdake commented:
@hkshaw1990 
```
Sure, I will do that.

Also, we thought a PR for istio/proxy and envoyproxy/envoy would be ready by now but we are consistently hitting new issues on ppc64le which we are trying to fix (see: discussion). So, the envoy PRs are not ready yet. We have been using an older version of envoy to test on our system with istio. Shall we still move ahead with this PR or wait till we get to the envoy PRs?
```

There are significant changes to move to envoyv2 in pilot already merged.  I'm not sure how long an older version of envoy would work or if it even works today.

I'd prefer @ldemailly or @costinm or @ZackButcher  answer this question as I'm unclear on the status and I may provide an incorrect answer.

Cheers
-steve

gyliu513 commented:
@hkshaw1990 can you help continue the work here?

hkshaw1990 commented:
Hi @gyliu513 its taking some time to get the envoy PR for `ppc64le` ready, since, there are some toolchain issues involved for upstreaming. @clnperez is working on that. As soon as the PR for envoy gets ready, I will rebase it to the latest master, since, this PR assumes that the `envoy` binary is available at the appropriate link.
